javascript/core/src/steps/types.ts

```diff
@@ -0,0 +1,100 @@
+/*!
+ * Copyright (c) Microsoft. All rights reserved.
+ * Licensed under the MIT license. See LICENSE file in the project.
+ */
+import type {
+	AggregateArgs,
+	BinArgs,
+	BinarizeArgs,
+	BooleanArgs,
+	ConvertArgs,
+	DedupeArgs,
+	DeriveArgs,
+	EraseArgs,
+	FetchArgs,
+	FillArgs,
+	FilterArgs,
+	FoldArgs,
+	GroupbyArgs,
+	ImputeArgs,
+	InputColumnListArgs,
+	JoinArgs,
+	LookupArgs,
+	MergeArgs,
+	OrderbyArgs,
+	PivotArgs,
+	RecodeArgs,
+	RenameArgs,
+	RollupArgs,
+	SampleArgs,
+	SelectArgs,
+	SetOperationArgs,
+	SpreadArgs,
+	UnfoldArgs,
+	UnrollArgs,
+	Verb,
+	WindowArgs,
+} from '../verbs/types/index.js'
+
+export interface Step<T = unknown> {
+	/**
+	 * A unique identifier for this step
+	 */
+	id: string
+
+	/**
+	 * helpful for documentation in JSON specs
+	 */
+	description?: string
+
+	/**
+	 * The verb being executed
+	 */
+	verb: Verb
+
+	/**
+	 * The verb arguments
+	 */
+	args: T
+
+	/**
+	 * The bound inputs
+	 */
+	inputs: Record<string, { node: string; output?: string }>
+
+	/**
+	 * The observed outputs to record.
+	 */
+	pinnedOutputs: Array<{ output?: string; name: string }>
```

Comment:
For serialized pipelines it would be nice if the inputs/outputs on the Step were symmetrically named and structured

Reply:
We could improve symmetry by using `outputs` instead of `pinnedOutputs`, but the structure won't match entirely.  Full symmetry is a bit of a misnomer because it's not a port-binding (like in the inputs area) but a publication to the TableStore. 

So inputs would be `Record<string, Binding>` and outputs would be `Record<string,string>`

Reply:
In my latest push this is a record<string,string> I'm not sure it's an improvement, but it rhymes with the inputs object. One caveat is that if you're using the default node output you have to use the key `default`: 

e.g. 
```js
{
   outputs: { default: 'table-1' }
}
```

I also updated the graph composer so that it will implicitly use the default output of the previous step as the default input if no inputs are defined. It feels a little weird, but it may be okay for this use case. Since the UI is going to be constructing these steps I'm not sure it's worth removing that explicit wiring.